File: src/store/dedup.ts
/**
 * 去重流程的状态。
 *
 * 和压缩那条线彻底分开：去重不挂在任何 job 上（D-102），也不共用队列。
 *
 * ```text
 * idle ──start()──> scanning ──有结果──> review ──apply()──> applying ──> done
 *   ^                   │                   │                              │
 *   └── cancel() ───────┘                   └── discard() ──> idle <───────┘
 * ```
 *
 * `review` 是唯一一屏能改东西的地方，也是唯一一道闸：**从 `review` 到
 * `applying` 的那次点击是整个应用里唯一不可撤销的操作**（文件进回收站，
 * 还能捞，但队列已经动过了）。所以勾选状态一律以库为准——每次改都立刻落库，
 * 应用被强杀也不丢；确认框上的两个数也一律问后端，不拿已加载的那几页去数。
 */
import { create } from "zustand";

import {
  ipc,
  onDedup,
  onDedupApply,
  toIpcError,
  type ApplyProgress,
  type ApplySummary,
  type DedupMode,
  type DedupProgress,
  type DedupReport,
  type DedupRun,
  type IpcError,
  type PendingRemovals,
  type Policy,
  type StoredGroup,
} from "@/lib/ipc";

export type DedupPhase = "idle" | "scanning" | "review" | "applying" | "done";

/** 一页多少组。一组通常 2~3 条，100 组约等于两三百行，翻一屏够用。 */
const PAGE = 100;

const NOTHING_PENDING: PendingRemovals = { count: 0, bytes: 0 };

/**
 * 相似度滑杆的三个数，单位是 256 位指纹上的汉明距离。
 *
 * 全部由基准 23 在真实照片上标定（`dedup/perceptual.rs`，后端还会再夹一道）：
 *
 * - `default` 16：缩放/重编码/提亮这类「同一张图的另一个版本」实测最多差 5 位，
 *   留了 3 倍余量；离实测的假配对最小值 62 还差 3.9 倍。
 * - `min` 4：再严也没意义，上面那几类一位都不会漏。
 * - `max` 56：刚够到「裁掉一圈边」那一类（实测最大 54），且仍在 62 之下。
 *   推到头也不该出现两张不相干的照片成组——那正是 ADR-031 修掉的毛病。
 * - `step` 4：256 位刻度上一位一格没有意义，14 格好拖。
 */
export const THRESHOLD = { min: 4, max: 56, step: 4, default: 16 } as const;

interface DedupState {
  phase: DedupPhase;
  roots: string[];
  mode: DedupMode;
  /** 感知模式的汉明距离阈值（256 位指纹上的位数）。精确模式不用。 */
  threshold: number;

  progress: DedupProgress | null;
  report: DedupReport | null;
  run: DedupRun | null;
  groups: StoredGroup[];
  /** 后面还有没有更多组。 */
  more: boolean;
  /** 当前用的保留策略。库里不存这个，它只描述「上一次是怎么勾的」。 */
  policy: Policy;
  /** 整个 run 勾了多少要删。**确认框上的数字只能用它。** */
  pending: PendingRemovals;

  applying: ApplyProgress | null;
  summary: ApplySummary | null;
  error: IpcError | null;

  addRoots: (roots: string[]) => void;
  removeRoot: (root: string) => void;
  setMode: (mode: DedupMode) => void;
  setThreshold: (n: number) => void;

  start: () => Promise<void>;
  cancel: () => Promise<void>;
  /** 启动时把上次没看完的结果摆回来。没有就什么都不做。 */
  resume: () => Promise<void>;
  loadMore: () => Promise<void>;
  toggleKeep: (memberId: number, keep: boolean) => Promise<void>;
  /** 只留这一条，同组其余的全勾成「删掉」。并排对比里挑完就是这一步。 */
  keepOnly: (groupId: number, memberId: number) => Promise<void>;
  choosePolicy: (policy: Policy) => Promise<void>;
  apply: () => Promise<void>;
  discard: () => Promise<void>;
  dismissError: () => void;
  /** 回到选目录那一屏，保留已选的根目录。 */
  reset: () => void;
}

/**
 * 当前订阅的退订函数。
 *
 * 放在 store 外面，理由同 {@link useScan}：它不是渲染要用的数据。
 */
let unlisten: (() => void) | null = null;

function stopListening() {
  unlisten?.();
  unlisten = null;
}

export const useDedup = create<DedupState>((set, get) => ({
  phase: "idle",
  roots: [],
  mode: "exact",
  // 这几个数都由基准 23 标定，后端 `clamp_threshold` 才是权威，这里只是初值。
  // 改之前先跑 `bench_perceptual_calibration`：它会红。
  threshold: THRESHOLD.default,

  progress: null,
  report: null,
  run: null,
  groups: [],
  more: false,
  policy: "manual",
  pending: NOTHING_PENDING,

  applying: null,
  summary: null,
  error: null,

  addRoots: (incoming) => set((s) => ({ roots: [...new Set([...s.roots, ...incoming])] })),
  removeRoot: (root) => set((s) => ({ roots: s.roots.filter((r) => r !== root) })),
  setMode: (mode) => set({ mode }),
  setThreshold: (threshold) => set({ threshold }),

  start: async () => {
    const { roots, mode, threshold, phase } = get();
    if (roots.length === 0) return;
    // 已经在跑就忽略。少了这道闸，第二次调用会先掐掉第一次的监听，再被后端
    // 以「已有查重在进行中」拒掉——查重照跑，报告却没人接（同 useScan）。
    if (phase === "scanning" || phase === "applying") return;

    stopListening();
    set({
      phase: "scanning",
      error: null,
      progress: null,
      report: null,
      run: null,
      groups: [],
      summary: null,
      pending: NOTHING_PENDING,
    });

    try {
      unlisten = onDedup({
        progress: (progress) => set({ progress }),
        report: (report) => {
          stopListening();
          void finish(report, set);
        },
      });
      await ipc.dedupStart(mode, roots, mode === "perceptual" ? threshold : null);
    } catch (e) {
      stopListening();
      set({ phase: "idle", error: toIpcError(e) });
    }
  },

  cancel: async () => {
    // 不在这里切状态：后端取消后会把 run 整个删掉再发一条 cancelled 的报告，
    // 收到它才知道确实收摊了。
    try {
      await ipc.dedupCancel();
    } catch (e) {
      stopListening();
      set({ phase: "idle", error: toIpcError(e) });
    }
  },

  resume: async () => {
    if (get().phase !== "idle") return;
    try {
      const run = await ipc.dedupLatest();
      // `applying` 也算没看完：上次删到一半被强杀，已经进回收站的带着 disposal
      // 标记，重新执行会自动跳过它们。
      if (!run || (run.status !== "ready" && run.status !== "applying")) return;
      const [groups, pending] = await Promise.all([
        ipc.dedupGroups(run.id, PAGE, 0),
        ipc.dedupPending(run.id),
      ]);
      if (groups.length === 0) return;
      set({
        phase: "review",
        run,
        roots: run.roots,
        mode: run.mode === "perceptual" ? "perceptual" : "exact",
        threshold: run.threshold ?? get().threshold,
        groups,
        more: groups.length === PAGE,
        pending,
        // 捞回来的勾选是库里那一份，不是任何策略现算的。说成 manual 才诚实。
        policy: "manual",
      });
    } catch (e) {
      set({ error: toIpcError(e) });
    }
  },

  loadMore: async () => {
    const { run, groups, more } = get();
    if (!run || !more) return;
    try {
      const page = await ipc.dedupGroups(run.id, PAGE, groups.length);
      set((s) => ({ groups: [...s.groups, ...page], more: page.length === PAGE }));
    } catch (e) {
      set({ error: toIpcError(e) });
    }
  },

  toggleKeep: async (memberId, keep) => {
    const { run } = get();
    if (!run) return;
    // 先改本地再落库：勾选要跟手，等一次 IPC 往返再动会有肉眼可见的迟滞。
    set((s) => ({ groups: patchMember(s.groups, memberId, { keep }) }));
    try {
      await ipc.dedupSetKeep(memberId, keep);
      set({ pending: await ipc.dedupPending(run.id), policy: "manual" });
    } catch (e) {
      // 落库失败就把本地改回去，别让界面显示一个库里没有的状态。
      set((s) => ({ groups: patchMember(s.groups, memberId, { keep: !keep }), error: toIpcError(e) }));
    }
  },

  keepOnly: async (groupId, memberId) => {
    const { run, groups } = get();
    if (!run) return;
    const group = groups.find((g) => g.id === groupId);
    if (!group) return;
    // 已经处置过的不碰：它的文件早就不在原处了，再改勾选只会让界面说谎。
    const changes = group.members
      .filter((m) => m.disposal === null && m.keep !== (m.id === memberId))
      .map((m) => ({ id: m.id, keep: m.id === memberId }));
    if (changes.length === 0) return;

    set((s) => ({
      groups: changes.reduce((gs, c) => patchMember(gs, c.id, { keep: c.keep }), s.groups),
    }));
    try {
      // 串行落库：一组撑死几条，而并发写同一张表只会让失败时更难说清改到哪儿了。
      for (const c of changes) await ipc.dedupSetKeep(c.id, c.keep);
      set({ pending: await ipc.dedupPending(run.id), policy: "manual" });
    } catch (e) {
      // 中途失败时本地已经全改了、库里只改了一半。以库为准重读这一整页，
      // 别去猜断在哪一条上——勾选状态说错就等于把删除范围说错。
      try {
        const fresh = await ipc.dedupGroups(run.id, Math.max(PAGE, groups.length), 0);
        set({ groups: fresh, pending: await ipc.dedupPending(run.id) });
      } catch {
        /* 重读也失败就只剩下报错，下面那行会说 */
      }
      set({ error: toIpcError(e) });
    }
  },

  choosePolicy: async (policy) => {
    const { run } = get();
    if (!run) return;
    try {
      await ipc.dedupApplyPolicy(run.id, policy);
      // 策略是整个 run 的重算，已加载的每一页都可能变，只能重读。
      const groups = await ipc.dedupGroups(run.id, Math.max(PAGE, get().groups.length), 0);
      set({ policy, groups, pending: await ipc.dedupPending(run.id) });
    } catch (e) {
      set({ error: toIpcError(e) });
    }
  },

  apply: async () => {
    const { run, phase } = get();
    if (!run || phase !== "review") return;

    stopListening();
    set({ phase: "applying", applying: null, summary: null, error: null });
    try {
      unlisten = onDedupApply({
        progress: (applying) => set({ applying }),
        applied: (summary) => {
          stopListening();
          set({ summary, phase: "done" });
          // 重读第一页，让已经进回收站的那些当场划掉。
          void ipc
            .dedupGroups(run.id, PAGE, 0)
            .then((groups) => set({ groups, more: groups.length === PAGE }))
            .catch(() => {});
        },
      });
      await ipc.dedupApply(run.id);
    } catch (e) {
      stopListening();
      set({ phase: "review", error: toIpcError(e) });
    }
  },

  discard: async () => {
    const { run } = get();
    if (run) {
      try {
        await ipc.dedupDiscard(run.id);
      } catch (e) {
        set({ error: toIpcError(e) });
        return;
      }
    }
    get().reset();
  },

  dismissError: () => set({ error: null }),

  reset: () => {
    stopListening();
    set({
      phase: "idle",
      progress: null,
      report: null,
      run: null,
      groups: [],
      more: false,
      policy: "manual",
      pending: NOTHING_PENDING,
      applying: null,
      summary: null,
      error: null,
    });
  },
}));

/**
 * 收到查重报告之后的收尾。
 *
 * 取消的那一份**不会**留在库里（后端直接删掉了）——半份结果看起来和完整结果
 * 一模一样，用户照着它删文件，真正的副本可能就在没扫到的那一半里。
 */
async function finish(report: DedupReport, set: (partial: Partial<DedupState>) => void) {
  if (report.cancelled) {
    set({ phase: "idle", progress: null, report: null });
    return;
  }
  set({ report });
  try {
    const run = await ipc.dedupLatest();
    if (!run || report.groups === 0) {
      // 一组都没有也是个结论，摆在 review 那一屏说清楚，不要退回选目录——
      // 那会让人以为自己压根没点到按钮。
      set({ phase: "review", run, groups: [], more: false, pending: NOTHING_PENDING });
      return;
    }
    const [groups, pending] = await Promise.all([
      ipc.dedupGroups(run.id, PAGE, 0),
      ipc.dedupPending(run.id),
    ]);
    set({
      phase: "review",
      run,
      groups,
      more: groups.length === PAGE,
      pending,
      // 和后端 `dedup_session::run` 落库时用的策略保持一致：精确组已按
      // 「路径最浅的留下」勾好，感知组一条都没勾（D-113）。
      policy: run.mode === "perceptual" ? "manual" : "shallowest_path",
    });
  } catch (e) {
    set({ phase: "idle", error: toIpcError(e) });
  }
}

/** 就地改一条成员，返回新的 groups。找不到就原样返回。 */
function patchMember(
  groups: StoredGroup[],
  memberId: number,
  patch: { keep: boolean },
): StoredGroup[] {
  return groups.map((g) =>
    g.members.some((m) => m.id === memberId)
      ? { ...g, members: g.members.map((m) => (m.id === memberId ? { ...m, ...patch } : m)) }
      : g,
  );
}

File: src-tauri/src/commands/dedup.rs
//! 去重相关的 IPC。
//!
//! 和扫描同构：耗时的两件事（查重、删除）**立刻返回**，进度靠事件推，
//! 否则 invoke 会挂住前端的 promise，用户连取消按钮都点不到。
//!
//! | 事件 | 载荷 | 时机 |
//! |---|---|---|
//! | `dedup://progress` | [`crate::core::dedup_session::DedupProgress`] | 查重过程中，已节流 |
//! | `dedup://report` | [`crate::core::dedup_session::DedupReport`] | 查重收尾一次 |
//! | `dedup://apply` | [`ApplyProgress`] | 删除过程中，已节流 |
//! | `dedup://applied` | [`ApplySummary`] | 删除收尾一次 |
//!
//! **查重和删除是两个命令，中间必须隔一次用户确认。** 不做成一个「一键清理」：
//! 那等于让一次点击同时授权「找」和「删」，而找出来的东西用户根本没看过。

use std::path::PathBuf;
use std::time::{Duration, Instant};

use serde::Serialize;
use tauri::{AppHandle, Emitter, Manager};
use ts_rs::TS;

use crate::core::dedup_session::{self, DedupMode, DedupProgress, DedupReport};
use crate::dedup::apply::{self, Outcome};
use crate::dedup::keep::Policy;
use crate::error::{Result, ZzError};
use crate::store::{DedupRun, StoredGroup};

use super::AppState;

pub const EVENT_PROGRESS: &str = "dedup://progress";
pub const EVENT_REPORT: &str = "dedup://report";
pub const EVENT_APPLY: &str = "dedup://apply";
pub const EVENT_APPLIED: &str = "dedup://applied";

/// 事件节流间隔。十万文件的查重每秒能产生上万条进度，不节流的话
/// WebView 光处理事件就卡住了（R10）。
const THROTTLE: Duration = Duration::from_millis(100);

#[derive(Debug, Clone, Copy, Serialize, TS)]
#[ts(export, export_to = "../../src/lib/bindings/")]
pub struct ApplyProgress {
    pub done: usize,
    pub total: usize,
    #[ts(type = "number")]
    pub reclaimed: u64,
}

/// 删完之后的交代。三个数分开报，因为它们对用户的含义完全不同。
#[derive(Debug, Clone, Default, Serialize, TS)]
#[ts(export, export_to = "../../src/lib/bindings/")]
pub struct ApplySummary {
    /// 进了回收站的条数。
    pub trashed: usize,
    /// 被安全机制挡下的条数（文件改过了、整组会被删空……）。**不是错误。**
    pub skipped: usize,
    /// 想删没删掉的。
    pub failed: usize,
    #[ts(type = "number")]
    pub reclaimed: u64,
    /// 头几条被跳过或失败的原因，给界面直接显示。全列出来在十万规模下没法看。
    pub notes: Vec<String>,
}

/// 当前勾选待删的量。确认框上要说的就是这两个数。
#[derive(Debug, Clone, Copy, Serialize, TS)]
#[ts(export, export_to = "../../src/lib/bindings/")]
pub struct PendingRemovals {
    pub count: usize,
    #[ts(type = "number")]
    pub bytes: u64,
}

/// 最多往 `notes` 里塞几条。
const MAX_NOTES: usize = 20;

/// 把界面传来的阈值夹回标定过的范围里（基准 23）。
///
/// 滑杆的两端在前端也写了一份，但那份只是遥控器：常量哪天飘了、或者有人直接调
/// 这个命令，都不该有办法把分组赶进噪声区——[`MAX_DISTANCE`] 之上就是实测的
/// 假配对区间，越过去就会把毫不相干的照片凑成一组（ADR-031）。
fn clamp_threshold(v: Option<u32>) -> u32 {
    use crate::dedup::perceptual::{DEFAULT_MAX_DISTANCE, MAX_DISTANCE, MIN_DISTANCE};
    v.unwrap_or(DEFAULT_MAX_DISTANCE).clamp(MIN_DISTANCE, MAX_DISTANCE)
}

/// 开始查重。立刻返回，进度走事件。
#[tauri::command]
pub fn dedup_start(
    app: AppHandle,
    state: tauri::State<'_, AppState>,
    mode: DedupMode,
    roots: Vec<PathBuf>,
    threshold: Option<u32>,
) -> Result<()> {
    if roots.is_empty() {
        return Err(ZzError::BadConfig("没有选择要查重的目录".into()));
    }
    let Some(cancel) = state.dedup.lock().expect("去重锁中毒").claim() else {
        return Err(ZzError::Other("已有查重在进行中".into()));
    };

    let db = state.db.clone();
    let threshold = clamp_threshold(threshold);
    // 用阻塞线程而不是 async 任务：查重全是同步的 CPU/IO，丢进 async 运行时
    // 会把它的工作线程占死，其他命令跟着一起卡住。
    std::thread::spawn(move || {
        let throttle = Throttle::new();
        let report = dedup_session::run(&db, mode, roots, threshold, &cancel, |p: DedupProgress| {
            // 收尾那一步（Saving）必须发出去，它之后就没有进度了。
            if matches!(p, DedupProgress::Saving) || throttle.ready() {
                let _ = app.emit(EVENT_PROGRESS, p);
            }
        });
        // 先腾位再发报告：用户看到报告就可能马上再查一轮。
        app.state::<AppState>().dedup.lock().expect("去重锁中毒").release(&cancel);
        match report {
            Ok(r) => {
                let _ = app.emit(EVENT_REPORT, r);
            }
            Err(e) => {
                tracing::error!(%e, "查重失败");
                let _ = app.emit(EVENT_REPORT, DedupReport { errors: 1, ..Default::default() });
            }
        }
    });
    Ok(())
}

/// 取消查重。已经算过的哈希留在缓存里，下次接着用。
#[tauri::command]
pub fn dedup_cancel(state: tauri::State<'_, AppState>) {
    if state.dedup.lock().expect("去重锁中毒").cancel() {
        tracing::info!("查重已取消");
    }
}

/// 最近一次查重。应用启动时用它决定要不要把上次没看完的结果摆出来。
#[tauri::command]
pub fn dedup_latest(state: tauri::State<'_, AppState>) -> Result<Option<DedupRun>> {
    state.db.latest_dedup_run()
}

/// 翻页读分组。
#[tauri::command]
pub fn dedup_groups(
    state: tauri::State<'_, AppState>,
    run_id: i64,
    limit: usize,
    offset: usize,
) -> Result<Vec<StoredGroup>> {
    state.db.list_dedup_groups(run_id, limit, offset)
}

/// 勾一条。直接落库，应用被关掉勾选不丢。
#[tauri::command]
pub fn dedup_set_keep(
    state: tauri::State<'_, AppState>,
    member_id: i64,
    keep: bool,
) -> Result<()> {
    state.db.set_member_keep(member_id, keep)
}

/// 当前勾了多少条要删、共多少字节。
///
/// 界面是翻页读的，自己数只能数到已加载的那几页；而删除作用于整个 run。
#[tauri::command]
pub fn dedup_pending(state: tauri::State<'_, AppState>, run_id: i64) -> Result<PendingRemovals> {
    let (count, bytes) = state.db.pending_removals(run_id)?;
    Ok(PendingRemovals { count, bytes })
}

/// 按策略重新勾一遍整个 run，返回被勾选删除的条数。
#[tauri::command]
pub fn dedup_apply_policy(
    state: tauri::State<'_, AppState>,
    run_id: i64,
    policy: Policy,
) -> Result<usize> {
    state.db.apply_keep_policy(run_id, policy)
}

/// 丢掉一次查重的结果。哈希缓存不动——那描述的是文件，不是这次的结论。
#[tauri::command]
pub fn dedup_discard(state: tauri::State<'_, AppState>, run_id: i64) -> Result<()> {
    state.db.delete_dedup_run(run_id)
}

/// 执行删除：把勾掉的送进回收站。立刻返回，进度走事件。
///
/// 这是**用户确认之后**才该调到的命令。安全判据（一组不能删空、文件改过就跳过）
/// 在 [`crate::dedup::apply`] 里，这一层不重复实现，也不绕过。
#[tauri::command]
pub fn dedup_apply(app: AppHandle, state: tauri::State<'_, AppState>, run_id: i64) -> Result<()> {
    let Some(cancel) = state.dedup.lock().expect("去重锁中毒").claim() else {
        return Err(ZzError::Other("查重还没结束".into()));
    };

    let db = state.db.clone();
    // 这两步会失败，而位子已经占上了：不还回去，用户就再也删不了第二批。
    let prepared = db.dedup_plans(run_id).and_then(|plans| {
        db.set_dedup_run_status(run_id, "applying")?;
        Ok(plans)
    });
    let plans = match prepared {
        Ok(plans) => plans,
        Err(e) => {
            state.dedup.lock().expect("去重锁中毒").release(&cancel);
            return Err(e);
        }
    };

    std::thread::spawn(move || {
        let throttle = Throttle::new();
        let results = apply::apply(&plans, &cancel, |p| {
            if p.done == p.total || throttle.ready() {
                let _ = app.emit(
                    EVENT_APPLY,
                    ApplyProgress { done: p.done, total: p.total, reclaimed: p.reclaimed },
                );
            }
        });

        // 先删后记（见 store::dedup::mark_member_disposed）。写库失败最多是界面
        // 少个标记，反过来会让用户以为文件还在。
        if let Err(e) = db.record_disposals(&results) {
            tracing::error!(%e, "删除结果落库失败，文件已经删了");
        }
        let _ = db.set_dedup_run_status(run_id, "done");
        app.state::<AppState>().dedup.lock().expect("去重锁中毒").release(&cancel);
        let _ = app.emit(EVENT_APPLIED, summarize(&results, &plans));
    });
    Ok(())
}

/// 把逐条结果折成一份交代。
fn summarize(results: &[(i64, Outcome)], plans: &[apply::GroupPlan]) -> ApplySummary {
    let size_of: std::collections::HashMap<i64, u64> =
        plans.iter().flat_map(|p| p.remove.iter().map(|t| (t.member_id, t.size))).collect();

    let mut s = ApplySummary::default();
    for (id, outcome) in results {
        match outcome {
            Outcome::Trashed => {
                s.trashed += 1;
                s.reclaimed += size_of.get(id).copied().unwrap_or(0);
            }
            Outcome::Skipped(why) => {
                s.skipped += 1;
                push_note(&mut s.notes, why);
            }
            Outcome::Failed(why) => {
                s.failed += 1;
                push_note(&mut s.notes, why);
            }
        }
    }
    s
}

/// 同一个原因只记一次——一组被跳过会产生成百条一模一样的说明。
fn push_note(notes: &mut Vec<String>, why: &str) {
    if notes.len() < MAX_NOTES && !notes.iter().any(|n| n == why) {
        notes.push(why.to_string());
    }
}

/// 时间节流。事件发得比 WebView 处理得快，界面就会卡住不动（R10）。
struct Throttle(std::sync::Mutex<Instant>);

impl Throttle {
    fn new() -> Self {
        // 减一个间隔，让第一条进度立刻发出去——否则界面开头有 100 ms 是空的。
        Self(std::sync::Mutex::new(Instant::now() - THROTTLE))
    }

    fn ready(&self) -> bool {
        let mut last = self.0.lock().expect("节流锁中毒");
        if last.elapsed() >= THROTTLE {
            *last = Instant::now();
            return true;
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::dedup::apply::{GroupPlan, Target};

    #[test]
    fn event_names_are_namespaced() {
        // 前端按前缀订阅，改名等于悄悄断掉界面更新。
        for e in [EVENT_PROGRESS, EVENT_REPORT, EVENT_APPLY, EVENT_APPLIED] {
            assert!(e.starts_with("dedup://"), "{e}");
        }
        assert_eq!(
            [EVENT_PROGRESS, EVENT_REPORT, EVENT_APPLY, EVENT_APPLIED]
                .iter()
                .collect::<std::collections::HashSet<_>>()
                .len(),
            4,
            "四个事件名不能撞"
        );
    }

    fn target(id: i64, size: u64) -> Target {
        Target { member_id: id, path: PathBuf::from(format!("/x/{id}")), size, mtime: 0 }
    }

    #[test]
    fn the_summary_counts_only_what_really_went_away() {
        // 「省下了多少」这个数字要给用户看，跳过和失败的不能算进去。
        let plans = [GroupPlan {
            group_id: 1,
            keep: vec![PathBuf::from("/x/keep")],
            remove: vec![target(1, 100), target(2, 200), target(3, 400)],
        }];
        let results = vec![
            (1, Outcome::Trashed),
            (2, Outcome::Skipped("文件已被修改，重复的结论不再成立")),
            (3, Outcome::Failed("权限不足".into())),
        ];
        let s = summarize(&results, &plans);
        assert_eq!((s.trashed, s.skipped, s.failed), (1, 1, 1));
        assert_eq!(s.reclaimed, 100);
        assert_eq!(s.notes.len(), 2, "跳过和失败各一条原因");
    }

    #[test]
    fn identical_reasons_are_not_repeated() {
        // 一组被整个跳过会产生成百条一模一样的说明，全塞进去没法看。
        let plans = [GroupPlan { group_id: 1, keep: vec![], remove: vec![] }];
        let results: Vec<_> =
            (0..500).map(|i| (i, Outcome::Skipped("整组都被勾选删除，已跳过"))).collect();
        let s = summarize(&results, &plans);
        assert_eq!(s.skipped, 500);
        assert_eq!(s.notes, ["整组都被勾选删除，已跳过"]);
    }

    #[test]
    fn notes_are_capped() {
        let plans = [GroupPlan { group_id: 1, keep: vec![], remove: vec![] }];
        let results: Vec<_> =
            (0..100).map(|i| (i, Outcome::Failed(format!("第 {i} 种错")))).collect();
        assert_eq!(summarize(&results, &plans).notes.len(), MAX_NOTES);
    }

    #[test]
    fn the_first_progress_event_is_not_swallowed() {
        // 节流器如果从「现在」起算，界面开头会有 100 ms 完全没反应。
        let t = Throttle::new();
        assert!(t.ready(), "第一条必须立刻放行");
        assert!(!t.ready(), "紧接着的要被拦住");
    }

    #[test]
    fn a_threshold_from_the_ui_cannot_reach_the_noise_floor() {
        use crate::dedup::perceptual::{DEFAULT_MAX_DISTANCE, MAX_DISTANCE, MIN_DISTANCE};
        assert_eq!(clamp_threshold(None), DEFAULT_MAX_DISTANCE, "没给就用标定过的默认值");
        assert_eq!(clamp_threshold(Some(DEFAULT_MAX_DISTANCE)), DEFAULT_MAX_DISTANCE);
        // 越界的一律夹回来。往上越界最要命：MAX_DISTANCE 之上就是实测的假配对区间。
        assert_eq!(clamp_threshold(Some(0)), MIN_DISTANCE);
        assert_eq!(clamp_threshold(Some(u32::MAX)), MAX_DISTANCE);
    }

    #[test]
    fn types_crossing_the_boundary_are_serializable() {
        let p = serde_json::to_value(DedupProgress::Walking { found: 7 }).unwrap();
        assert_eq!(p["stage"], "walking", "带 tag 的枚举，前端靠它分支");
        assert_eq!(p["found"], 7);
        let r = serde_json::to_value(DedupReport::default()).unwrap();
        assert_eq!(r["groups"], 0);
        assert_eq!(serde_json::to_value(DedupMode::Perceptual).unwrap(), "perceptual");
        assert_eq!(serde_json::to_value(Policy::ShallowestPath).unwrap(), "shallowest_path");
    }
}
